snuba/subscriptions/data.py

```diff
@@ -328,10 +336,57 @@ def build_request(
         timestamp: datetime,
         offset: Optional[int],
         timer: Timer,
+        metrics: Optional[MetricsBackend] = None,
     ) -> Request:
-        # TODO: Add a flag to use the SnQL data if it exists.
-        data = self.to_legacy()
-        return data.build_request(dataset, timestamp, offset, timer)
+        try:
+            if metrics is not None:
+                metrics.increment(
+                    "snql.subscription.delegate.incoming",
+                    tags={"project_id": str(self.project_id)},
+                )
+            snql_rollout_pct = state.get_config("snql_subscription_rollout_pct", 0.0)
+            assert isinstance(snql_rollout_pct, float)
+            snql_rollout_projects_raw = state.get_config(
+                "snql_subscription_rollout_projects", ""
+            )
+            assert isinstance(snql_rollout_projects_raw, str)
+            snql_rollout_projects = set(
+                [int(s.strip()) for s in snql_rollout_projects_raw.split(",")]
+            )
+            use_snql = (
+                snql_rollout_pct > 0.0
+                and self.project_id in snql_rollout_projects
+                and random.random() <= snql_rollout_pct
+            )
+            if use_snql:
+                if metrics is not None:
+                    metrics.increment(
+                        "snql.subscription.delegate.use_snql",
+                        tags={"project_id": str(self.project_id)},
+                    )
+                return self.to_snql().build_request(dataset, timestamp, offset, timer)
```

Comment:
This catches issue that can happen when generating the request but not issues that can happen when running the query on clickhouse. What about catching the exception here if we are executing a delegate subscription in snql mode and retrying by forcing the legacy mode?
https://github.com/getsentry/snuba/blob/master/snuba/subscriptions/worker.py#L84-L86

So the consumer would not get stuck till we turn snql off in case of issues.